File: src/main/GO/leetcode/lists/offer/Offer48.go
package offer

import "math"

/**
 * @author wujingxinit@outlook.com
 * @date 2023/1/4 22:36
 */

func lengthOfLongestSubstring(s string) int {

	if len(s) == 0 {
		return 0
	}
	mp := make(map[byte]int)
	j := -1
	max := math.MinInt32
	for i := 0; i < len(s); i++ {
		if v, ok := mp[s[i]]; ok {
			j = Max(v, mp[s[i]])
		}
		mp[s[i]] = i
		max = Max(max, i-j)
	}
	return max
}

File: java/src/leetcode/lists/offer/Offer48.java
package leetcode.lists.offer;

import java.util.HashMap;

/**
 * @author wujingxinit@outlook.com
 * @date 2023/9/13 23:24
 */
public class Offer48 {

    public int lengthOfLongestSubstring(String s) {
        if (s == null || s.isEmpty()) {
            return 0;
        }
        int left = 0;
        int right = 0;
        int res = 0;
        HashMap<Character, Integer> map = new HashMap<>();
        while (right < s.length()) {
            char c = s.charAt(right);
            map.put(c, map.getOrDefault(c, 0) + 1);
            while (map.get(c) > 1) {
                char d = s.charAt(left);
                map.put(d, map.getOrDefault(d, 0) - 1);
                left++;
            }
            res = Math.max(res, right - left + 1);
            right++;
        }
        return res;
    }
}
